granular_flow.go

```diff
@@ -41,227 +39,257 @@ func (c *baseConn) granularFlow(
-			buf.PopUint8()  // cardianlity
+			// num headers is always 0
+			// cardianlity
+			// input descriptor ID
+			r.Discard(19)
 
 			// input descriptor
-			buf.PopUUID()
-			descs.in = buf.PopBytes()
+			descs.in = r.PopBytes()
 
 			// output descriptor
-			buf.PopUUID()
-			descs.out = buf.PopBytes()
+			r.Discard(16) // descriptor ID
+			descs.out = r.PopBytes()
 		case message.ReadyForCommand:
-			buf.PopUint16() // header count (assume 0)
-			buf.PopUint8()  // transaction state
+			// header count (assume 0)
+			// transaction state
+			r.Discard(3)
+
+			done.Signal()
 		case message.ErrorResponse:
-			return descs, decodeError(buf)
+			err = wrapAll(err, decodeError(r))
+			done.Signal()
 		default:
-			return descs, fmt.Errorf(
-				"unexpected message type: 0x%x",
-				buf.MsgType,
-			)
+			if e := c.fallThrough(r); e != nil {
+				// the connection will not be usable after this x_x
+				return descPair{}, e
+			}
 		}
 	}
 
-	return descs, nil
+	if r.Err != nil {
+		return descPair{}, r.Err
+	}
+
+	return descs, err
 }
 
 func (c *baseConn) execute(
-	ctx context.Context,
+	r *buff.Reader,
 	out reflect.Value,
 	q query,
 	tp reflect.Type,
 	cdcs codecPair,
 ) error {
-	buf := buff.New(c.buffer[:0])
-	buf.BeginMessage(message.Execute)
-	buf.PushUint16(0)       // no headers
-	buf.PushBytes([]byte{}) // no statement name
-	cdcs.in.Encode(buf, q.args)
-	buf.EndMessage()
+	c.writer.BeginMessage(message.Execute)
+	c.writer.PushUint16(0)       // no headers
+	c.writer.PushBytes([]byte{}) // no statement name
+	cdcs.in.Encode(c.writer, q.args)
+	c.writer.EndMessage()
 
-	buf.BeginMessage(message.Sync)
-	buf.EndMessage()
+	c.writer.BeginMessage(message.Sync)
+	c.writer.EndMessage()
 
-	err := c.writeAndRead(ctx, buf.Unwrap())
-	if err != nil {
-		return err
+	if e := c.writer.Send(c.conn); e != nil {
+		return e
 	}
 
 	tmp := out
-	err = ErrZeroResults
-	for buf.Next() {
-		switch buf.MsgType {
+	err := ErrZeroResults
+	done := buff.NewSignal()
+
+	for r.Next(done.Chan) {
+		switch r.MsgType {
 		case message.Data:
-			buf.PopUint16() // number of data elements (always 1)
+			r.Discard(2) // number of data elements (always 1)
 
 			if !q.flat() {
 				val := reflect.New(tp).Elem()
-				cdcs.out.Decode(buf, unsafe.Pointer(val.UnsafeAddr()))
+				cdcs.out.Decode(r, unsafe.Pointer(val.UnsafeAddr()))
 				tmp = reflect.Append(tmp, val)
 			} else {
-				cdcs.out.Decode(buf, unsafe.Pointer(out.UnsafeAddr()))
+				cdcs.out.Decode(r, unsafe.Pointer(out.UnsafeAddr()))
 			}
 
-			err = nil
+			if err == ErrZeroResults {
+				err = nil
+			}
 		case message.CommandComplete:
-			buf.PopUint16() // header count (assume 0)
-			buf.PopBytes()  // command status
+			r.Discard(2) // header count (assume 0)
+			r.PopBytes() // command status
 		case message.ReadyForCommand:
-			buf.PopUint16() // header count (assume 0)
-			buf.PopUint8()  // transaction state
+			// header count (assume 0)
+			// transaction state
+			r.Discard(3)
```

Comment:
Need a helper for reading transaction state. The latest one should be stored in a private field on the connection object. We'll need to add public APIs to allow users to see it (not sure if we support that in JS/Python, but we should).

Reply:
:+1: created issue https://github.com/edgedb/edgedb-go/issues/43